Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Node.lean

```diff
@@ -0,0 +1,228 @@
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine.Point
+
+/-!
+
+# Singular cubics with nodes
+
+We develop the basic theory of singular cubics with nodes.
+
+## TODO (Andrew)
+- classify singular cubics with nodes
+
+-/
+
+namespace WeierstrassCurve
+
+variable {R K : Type*} [CommRing R] (E : WeierstrassCurve R) [Field K]
+
+/-- A Weierstrass curve is in split-singular normal form if it is of the form `y² + a₁xy = x³`.
+
+TODO(Andrew): Any singular cubic with a ractional point whose tangent splits in the base field
+can be turned into split-singular normal form. -/
+class IsSplitSingularNF (E : WeierstrassCurve R) where
+  a₂ (E) : E.a₂ = 0
+  a₃ (E) : E.a₃ = 0
+  a₄ (E) : E.a₄ = 0
+  a₆ (E) : E.a₆ = 0
+
+section
+
+variable [E.IsSplitSingularNF]
+
+@[simp] alias a₂_of_isSplitSingularNF := IsSplitSingularNF.a₂
+@[simp] alias a₃_of_isSplitSingularNF := IsSplitSingularNF.a₃
+@[simp] alias a₄_of_isSplitSingularNF := IsSplitSingularNF.a₄
+@[simp] alias a₆_of_isSplitSingularNF := IsSplitSingularNF.a₆
+        lemma b₂_of_isSplitSingularNF : E.b₂ = E.a₁ ^ 2   := by simp [b₂]
+@[simp] lemma b₄_of_isSplitSingularNF : E.b₄ = 0          := by simp [b₄]
+@[simp] lemma b₆_of_isSplitSingularNF : E.b₆ = 0          := by simp [b₆]
+@[simp] lemma b₈_of_isSplitSingularNF : E.b₈ = 0          := by simp [b₈]
+        lemma c₄_of_isSplitSingularNF : E.c₄ = E.a₁ ^ 4   := by simp [c₄, b₂]; ring
+        lemma c₆_of_isSplitSingularNF : E.c₆ = - E.a₁ ^ 6 := by simp [c₆, b₂]; ring
+@[simp] lemma Δ_of_isSplitSingularNF  : E.Δ  = 0          := by simp [Δ]
+
+end
+
+section IsNodal
+
+/-- We say that a `WeierstrassCurve` is nodal if its `Δ = 0` but its `c₄ ≠ 0`. -/
+class IsNodal (E : WeierstrassCurve R) : Prop where
+  c₄_eq_zero (E) : E.c₄ ≠ 0
+  Δ_eq_zero (E) : E.Δ = 0
+
+alias c₄_of_isNodal := IsNodal.c₄_eq_zero
+@[simp] alias Δ_of_isNodal := IsNodal.Δ_eq_zero
+
+lemma c₄_pow_three_of_isNodal [E.IsNodal] : E.c₄ ^ 3 = E.c₆ ^ 2 := by
+  simpa [sub_eq_zero, @eq_comm R 0] using E.c_relation
+
+lemma c₆_of_isNodal [IsReduced R] [E.IsNodal] : E.c₆ ≠ 0 :=
+  fun e ↦ E.c₄_of_isNodal (by simpa [e] using E.c₄_pow_three_of_isNodal)
+
+lemma a₁_of_isSplitSingularNF_of_isNodal [IsReduced R] [E.IsNodal] [E.IsSplitSingularNF] :
+    E.a₁ ≠ 0 := by
+  simpa [c₄_of_isSplitSingularNF] using E.c₄_of_isNodal
+
+lemma pointEquivOfIsNodalOfIsSplitSingularNF.aux [IsReduced R]
+    (E : WeierstrassCurve R) [E.IsSplitSingularNF]
+    (x y : R) (h : E.toAffine.Nonsingular x y) :
+    x ≠ 0 ∧ y ≠ -E.a₁ * x ∧ y ≠ 0 := by
+  have H₁ : y ^ 2 + E.a₁ * x * y - x ^ 3 = 0 := by simpa [Affine.equation_iff'] using h.1
+  by_contra H
+  simp only [ne_eq, neg_mul, not_and_or, not_not] at H
+  suffices x ^ 3 = 0 by aesop
+  obtain (rfl | rfl | rfl) := H
+  · simp
+  · linear_combination -H₁
+  · linear_combination -H₁
+
+lemma pointEquivOfIsNodalOfIsSplitSingularNF_nonsingular
+    (E : WeierstrassCurve K) [E.IsNodal] [E.IsSplitSingularNF] (k : K)
```

Comment:
Do you think it's a good idea to add an instance `IsSplitSingularNF -> IsNodal`? In this case here (and all the followings) you only need to write `[E.IsSplitSingularNF]`.